### PR TITLE
NIFI-1829 - Create new DebugFlow processor.

### PR DESCRIPTION
**Pull request by jskora**
* Create DebugFlow processor for use in testing and troubleshooting the processor framework.

* 0.x - commits as is.
* 1.x - in o.a.n.p.s.DebugFlow.java replace ProcessorLog references with ComponentLog.

### CODE REVIEW

**Comment by trkurc** [comment]
reviewing this

File: nifi-nar-bundles/nifi-standard-bundle/nifi-standard-processors/src/main/java/org/apache/nifi/processors/standard/DebugFlow.java [hunk 1]
```diff
@@ -0,0 +1,414 @@
...
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+
+@EventDriven()
+@Tags({"test", "debug", "processor", "utility", "flow", "flowfile"})
+@CapabilityDescription("This processor aids in the testing and debugging of the flowfile framework by allowing "
```
Comment (trkurc) [inline nifi-nar-bundles/nifi-standard-bundle/nifi-standard-processors/src/main/java/org/apache/nifi/processors/standard/DebugFlow.java]: recommend s/flowfile/FlowFile/ for consistency

File: nifi-nar-bundles/nifi-standard-bundle/nifi-standard-processors/src/main/java/org/apache/nifi/processors/standard/DebugFlow.java [hunk 2]
```diff
@@ -0,0 +1,414 @@
...
+        + "a developer or flow manager to force various responses to a flowfile.  In response to a receiving a flowfile "
+        + "it can route to the success or failure relationships, rollback, rollback and yield, rollback with penalty, "
+        + "or throw an exception.  In addition, if using a timer based scheduling strategy, upon being triggered without "
+        + "a flowfile, it can be configured to throw an exception,  when triggered without a flowfile.\n"
```
Comment (trkurc) [inline nifi-nar-bundles/nifi-standard-bundle/nifi-standard-processors/src/main/java/org/apache/nifi/processors/standard/DebugFlow.java]: I think the last "when triggered ..." is redundant

File: nifi-nar-bundles/nifi-standard-bundle/nifi-standard-processors/src/main/java/org/apache/nifi/processors/standard/DebugFlow.java [hunk 3]
```diff
@@ -0,0 +1,414 @@
...
+                        curr_ff_resp = curr_ff_resp.getNextCycle();
+                    }
+                }
+                if (curr_ff_resp == FlowfileResponse.FF_EXCEPTION_RESPONSE) {
+                    if (FF_EXCEPTION_CURR < FF_EXCEPTION_MAX) {
+                        FF_EXCEPTION_CURR += 1;
+                        logger.info("DebugFlow throwing NPE file={} UUID={}",
+                                new Object[]{ff.getAttribute(CoreAttributes.FILENAME.key()),
+                                        ff.getAttribute(CoreAttributes.UUID.key())});
+                        throw new NullPointerException("forced by " + this.getClass().getName());
```
Comment (trkurc) [inline nifi-nar-bundles/nifi-standard-bundle/nifi-standard-processors/src/main/java/org/apache/nifi/processors/standard/DebugFlow.java]: Why an NPE here? Would being allowed to specify the class of a Throwable in the configuration to throw here be overkill?

File: nifi-nar-bundles/nifi-standard-bundle/nifi-standard-processors/src/main/java/org/apache/nifi/processors/standard/DebugFlow.java [hunk 4]
```diff
@@ -0,0 +1,414 @@
...
+        + "  5. rollback the flowfile with penalty.\n"
+        + "  6. throw an exception.\n"
+        + "\n"
+        + "The order of responses when no flow file is received are:"
+        + "  1. yield the context.\n"
+        + "  2. throw an exception.\n"
+        + "  3. do nothing and return.\n"
+        + "\n"
+        + "By default, the processor is configured to perform each response one time.  After processing the list of "
+        + "responses it will resume from the top of the list.\n"
```
Comment (trkurc) [inline nifi-nar-bundles/nifi-standard-bundle/nifi-standard-processors/src/main/java/org/apache/nifi/processors/standard/DebugFlow.java]: looking at the code below, it looks like most are configured for 0 times.

File: nifi-nar-bundles/nifi-standard-bundle/nifi-standard-processors/src/main/java/org/apache/nifi/processors/standard/DebugFlow.java [hunk 5]
```diff
@@ -0,0 +1,414 @@
...
+    private Integer FF_ROLLBACK_MAX = 0;
+    private Integer FF_YIELD_MAX = 0;
+    private Integer FF_PENALTY_MAX = 0;
+    private Integer FF_EXCEPTION_MAX = 0;
+
+    private Integer NO_FF_EXCEPTION_MAX = 0;
+    private Integer NO_FF_YIELD_MAX = 0;
+    private Integer NO_FF_SKIP_MAX = 0;
+
+    private Integer FF_SUCCESS_CURR = 0;
```
Comment (trkurc) [inline nifi-nar-bundles/nifi-standard-bundle/nifi-standard-processors/src/main/java/org/apache/nifi/processors/standard/DebugFlow.java]: as these are not constants, they shouldn't be all caps. also, I believe there will be concurrency issues with this (need locking and safe publication)

File: nifi-nar-bundles/nifi-standard-bundle/nifi-standard-processors/src/main/java/org/apache/nifi/processors/standard/DebugFlow.java (hunk 2 above)
Comment (jskora) [inline nifi-nar-bundles/nifi-standard-bundle/nifi-standard-processors/src/main/java/org/apache/nifi/processors/standard/DebugFlow.java]: Agreed.

File: nifi-nar-bundles/nifi-standard-bundle/nifi-standard-processors/src/main/java/org/apache/nifi/processors/standard/DebugFlow.java (hunk 3 above)
Comment (jskora) [inline nifi-nar-bundles/nifi-standard-bundle/nifi-standard-processors/src/main/java/org/apache/nifi/processors/standard/DebugFlow.java]: I have broaden it out to any RuntimeException.  Other exceptions will be expected to be checked causing compilation errors.

File: nifi-nar-bundles/nifi-standard-bundle/nifi-standard-processors/src/main/java/org/apache/nifi/processors/standard/DebugFlow.java (hunk 5 above)
Comment (jskora) [inline nifi-nar-bundles/nifi-standard-bundle/nifi-standard-processors/src/main/java/org/apache/nifi/processors/standard/DebugFlow.java]: I will fix the names and looking into cleaning up the concurrency problems.

**Comment by joewitt** [comment]
The name isn't great but not sure what would be better.  It is indeed a debugger for the framework and less so for a flow.  It is perhaps good to call it 'CreateChaos'.  That said i personally think this is a great idea to have in the build and part of the standard processors.  It would only be used by power users or curious types.  And for everyone else it is a noop.  But it could make it easier for us to demonstrate the evil effects of constant exceptions in a flow and do interesting regression checks on various infrastructures.  There is admittedly probably a better place for this but with the upcoming nar registry work we can tackle that then.

**Comment by JPercivall** [comment]
@jskora rebased and addressed comments, @trkurc are you able to finish reviewing this for 0.7.0?

**Comment by trkurc** [comment]
Not until this weekend

**Comment by trkurc** [comment]
It looks like the issues I had were addressed, but I won't have a chance to dive in deeper until Saturday, maybe

**Comment by JPercivall** [comment]
@trkurc ok, in order to get 0.7.0 as soon as possible, would you mind if another committer took over reviewing?

**Comment by mosermw** [comment]
I will review.

**Comment by trkurc** [comment]
I don't mind even a little! Thanks @mosermw

File: nifi-nar-bundles/nifi-standard-bundle/nifi-standard-processors/src/main/java/org/apache/nifi/processors/standard/DebugFlow.java [hunk 6]
```diff
@@ -0,0 +1,499 @@
...
+        + "  2. transfer FlowFile to failure relationship.\n"
+        + "  3. rollback the FlowFile without penalty.\n"
+        + "  4. rollback the FlowFile and yield the context.\n"
+        + "  5. rollback the FlowFile with penalty.\n"
+        + "  6. throw an exception.\n"
+        + "\n"
+        + "Triggered without a FlowFile, the processor can produce the following responses."
+        + "  1. do nothing and return.\n"
+        + "  2. throw an exception.\n"
+        + "  3. yield the context.\n")
```
Comment (mosermw) [inline nifi-nar-bundles/nifi-standard-bundle/nifi-standard-processors/src/main/java/org/apache/nifi/processors/standard/DebugFlow.java]: Formatting with \n doesn't work inside a CapabilityDescription annotation.  It just gets smashed together.  If you are OK with that @jskora then we can leave it.

File: nifi-nar-bundles/nifi-standard-bundle/nifi-standard-processors/src/main/java/org/apache/nifi/processors/standard/DebugFlow.java [hunk 7]
```diff
@@ -0,0 +1,499 @@
...
+                    if (noFlowFileCurrException < noFlowFileMaxException) {
+                        noFlowFileCurrException += 1;
+                        logger.info("DebugFlow throwing NPE with no flow file");
+                        String message = "forced by " + this.getClass().getName();
+                        RuntimeException rte;
+                        try {
+                            rte = noFlowFileExceptionClass.getConstructor(String.class).newInstance(message);
+                            throw rte;
+                        } catch (InstantiationException | IllegalAccessException | InvocationTargetException | NoSuchMethodException e) {
+                            e.printStackTrace();
```
Comment (mosermw) [inline nifi-nar-bundles/nifi-standard-bundle/nifi-standard-processors/src/main/java/org/apache/nifi/processors/standard/DebugFlow.java]: Eeek e.printStackTrace() should be changed to use logger.

File: nifi-nar-bundles/nifi-standard-bundle/nifi-standard-processors/src/main/java/org/apache/nifi/processors/standard/DebugFlow.java (hunk 6 above)
Comment (mattyb149) [inline nifi-nar-bundles/nifi-standard-bundle/nifi-standard-processors/src/main/java/org/apache/nifi/processors/standard/DebugFlow.java]: I wonder if there's an abbreviated version of the doc for capability description, and move this out to an additional details page

Comment (jskora) [inline nifi-nar-bundles/nifi-standard-bundle/nifi-standard-processors/src/main/java/org/apache/nifi/processors/standard/DebugFlow.java]: Split out to additional details page.

File: nifi-nar-bundles/nifi-standard-bundle/nifi-standard-processors/src/main/java/org/apache/nifi/processors/standard/DebugFlow.java (hunk 7 above)
Comment (jskora) [inline nifi-nar-bundles/nifi-standard-bundle/nifi-standard-processors/src/main/java/org/apache/nifi/processors/standard/DebugFlow.java]: Validation will have verified that the class exists and can be loaded by the processor, so this should (haha) never occur.  I moved it to a debug level log message.

Comment (jskora) [inline nifi-nar-bundles/nifi-standard-bundle/nifi-standard-processors/src/main/java/org/apache/nifi/processors/standard/DebugFlow.java]: That is, I moved both occurrences.

**Comment by jskora** [comment]
Ok, I think that last commit covers the review feedback.

**Comment by mosermw** [comment]
Completed review, latest changes look good, unit testing and integration testing worked as expected, code passes -Pcontrib-check.

+1 I will squash commits and merge to 0.x and master shortly.  Thanks @jskora.

**Pull request closed without merging**